src/main/java/sorald/sonar/RuleVerifier.java

```diff
@@ -54,6 +83,78 @@ public static void verifyHasIssue(List<String> filesToScan, JavaFileScanner chec
+
+    private static InputFile toInputFile(File baseDir, String filename) {
+        // must append a separator to the basedir string as Sonar appends the filenames directly to
+        // it
+        final String baseDirStr = baseDir.toString() + File.separator;
+        try {
+            return new TestInputFileBuilder(baseDirStr, filename)
+                    .setContents(new String(Files.readAllBytes(Paths.get(filename)), UTF_8))
+                    .setCharset(UTF_8)
+                    .setLanguage("java")
+                    .build();
+        } catch (IOException e) {
+            throw new RuntimeException("failed to read file " + filename);
+        }
+    }
+
+    @SuppressWarnings("UnstableApiUsage")
+    private static JavaAstScanner createAstScanner(
+            SonarComponents sonarComponents, List<JavaFileScanner> checks) {
+        JavaAstScanner scanner = new JavaAstScanner(sonarComponents);
+        VisitorsBridge visitorsBridge =
+                new VisitorsBridge(
+                        checks,
+                        // TODO set the classpath to something reasonable
+                        Collections.emptyList(),
+                        sonarComponents,
+                        SymbolicExecutionMode.ENABLED);
+        // TODO we may want to set the version of the visitors bridge, as without setting it the
+        //      implementation defaults to the highest version supported by the parser (currently
+        //      Java 14)
+        //      visitorsBridge.setJavaVersion(new JavaVersionImpl(8));
+        scanner.setVisitorBridge(visitorsBridge);
+        return scanner;
+    }
+
+    private static SoraldSonarComponents createSonarComponents(File baseDir) {
+        // FIXME The SensorContextTester is an internal and unstable component in sonar,
+        //       we should implement our own SensorContext
+        SensorContextTester context = SensorContextTester.create(baseDir);
+        SoraldSonarComponents sonarComponents = new SoraldSonarComponents(context.fileSystem());
+        sonarComponents.setSensorContext(context);
+        return sonarComponents;
+    }
+
+    /**
+     * A simple subclass of SonarComponents that stores all analyzer messages. These are by default
+     * stored in a storage container, but it seems easier for our use case to just intercept them.
+     *
+     * <p>This IS a bit of a hack, so it wouldn't be unreasonable to try to do this the "proper
+     * way".
+     */
+    private static class SoraldSonarComponents extends SonarComponents {
+        private final List<AnalyzerMessage> messages;
+
+        public SoraldSonarComponents(DefaultFileSystem fs) {
+            // FIXME I'm very unsure about supplying null for all of these constructor values.
+            //       They should not be null, but at this time I don't know that to put there, and
+            //       with our current usage this hack appears to work.
+            super(null, fs, null, null, null, null);
```

Comment:
Like the comment says, this is a pretty nasty hack. It appears to work, but it's definitely not pretty nor safe as it initializes fields that should not be `null` to `null`.